### ISSUE TITLE
google_gke_hub_feature_membership for servicemesh doesn't support multiple clusters

### ISSUE THREAD

**Issue by goobysnack** [hashicorp/terraform-provider-google#15233]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version
Terraform v1.1.9
terragrunt version v0.38.8

### Affected Resource(s)

google_gke_hub_feature_membership (for servicemesh feature)

### Terraform Configuration Files

```tf
resource "google_gke_hub_feature_membership" "mesh" {
  count = var.join_fleet ? 1 : 0
  project = var.project_id
  location = "global"
  feature = "servicemesh"
  membership = google_gke_hub_membership.membership[0].id
  mesh {
    management = "MANAGEMENT_AUTOMATIC"
  }
  provider = google-beta
}
```

### Debug Output

```
Terraform will perform the following actions:

  # module.gcp_gkeclusters["cluster-1"].google_gke_hub_feature_membership.mesh[0] will be created
  + resource "google_gke_hub_feature_membership" "mesh" {
      + feature    = "servicemesh"
      + id         = (known after apply)
      + location   = "global"
      + membership = "projects/project-1/locations/global/memberships/cluster-1"
      + project    = "project-1"

      + mesh {
          + management = "MANAGEMENT_AUTOMATIC"
        }
    }

  # module.gcp_gkeclusters["cluster-2"].google_gke_hub_feature_membership.mesh[0] will be created
  + resource "google_gke_hub_feature_membership" "mesh" {
      + feature    = "servicemesh"
      + id         = (known after apply)
      + location   = "global"
      + membership = "projects/project-1/locations/global/memberships/cluster-2"
      + project    = "project-1"

      + mesh {
          + management = "MANAGEMENT_AUTOMATIC"
        }
    }

Plan: 2 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

module.gcp_gkeclusters["cluster-2"].google_gke_hub_feature_membership.mesh[0]: Creating...
module.gcp_gkeclusters["cluster-1"].google_gke_hub_feature_membership.mesh[0]: Creating...

Error: Error creating FeatureMembership: Resource already exists - apply blocked by lifecycle params: &beta.FeatureMembership{Mesh:(*beta.FeatureMembershipMesh)(0xc00215c9d8), Configmanagement:(*beta.FeatureMembershipConfigmanagement)(0x5089460), Project:(*string)(0xc002214310), Location:(*string)(0xc002214180), Feature:(*string)(0xc002214140), Membership:(*string)(0xc0022141c0)}.

  with module.gcp_gkeclusters["cluster-2"].google_gke_hub_feature_membership.mesh[0],
  on .terraform/modules/gcp_gkeclusters/main.tf line 162, in resource "google_gke_hub_feature_membership" "mesh":
 162: resource "google_gke_hub_feature_membership" "mesh" {


Error: Error creating FeatureMembership: Resource already exists - apply blocked by lifecycle params: &beta.FeatureMembership{Mesh:(*beta.FeatureMembershipMesh)(0xc0022728e8), Configmanagement:(*beta.FeatureMembershipConfigmanagement)(0x5089460), Project:(*string)(0xc0021673f0), Location:(*string)(0xc002167260), Feature:(*string)(0xc002167220), Membership:(*string)(0xc0021672a0)}.

  with module.gcp_gkeclusters["cluster-1"].google_gke_hub_feature_membership.mesh[0],
  on .terraform/modules/gcp_gkeclusters/main.tf line 162, in resource "google_gke_hub_feature_membership" "mesh":
 162: resource "google_gke_hub_feature_membership" "mesh" {

ERRO[0032] Terraform invocation failed in /tmp/terragrunt/aK1nXGfrTzvhopPiwyvUHvHBtaE/VsMQKd4b-YSkZgFIMgKoEwagiqQ  prefix=[/data/poc-1] 
ERRO[0032] 1 error occurred:
	* exit status 1
```

### Expected Behavior

The mesh membership config for each cluster should have updated.  It should work the same as these two separate gcloud commands:
```
gcloud container fleet mesh update --management automatic --project poc-1 --memberships cluster-1  --location global
Waiting for Feature Service Mesh to be updated...done.                                                                                                                                                                    

gcloud container fleet mesh update --management automatic --project poc-1 --memberships cluster-2  --location global
Waiting for Feature Service Mesh to be updated...done.  
```
### Actual Behavior
There's some conflict with the lifecycle in the config.  Each cluster should update independently, but are being treated as the same API object (I THINK?).

### Steps to Reproduce

1. `terraform apply`

### References

* #0000

b/293790010

**Comment by goobysnack** [comment]
Per the documentation, we can't use a list either (and without a Data block, how do we get a list of clusters?).  So how do we iterate over multiple clusters? 

https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/gke_hub_feature_membership#membership

**Comment by roaks3** [comment]
I believe this is a duplicate of https://github.com/hashicorp/terraform-provider-google/issues/10934. Unfortunately I'm not noticing a workaround there, just to use `gcloud` for now.

**Comment by goobysnack** [comment]
I think that the resource thinks that the mesh config is a single object where gcloud treats each registration membership separately.  If it's a single object to the state, I think the provider just needs to pivot to an update and not a creation.  I could be talking apples and oranges here.  Per the reference to the duplicate, I think that is referring to the member ID, and this isn't that situation.

**Comment by edwardmedia** [comment]
b/293790010

**Comment by djazayeri** [comment]
IIUC you need one google_gke_hub_feature_membership for each feature * membership. (So, 2 of these for the mesh feature with 2 clusters.)

Maybe there's a way to clarify this in the docs at https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/gke_hub_feature_membership#example-usage---service-mesh ?

**Comment by goobysnack** [comment]
@djazayeri I believe that's what I tried to do.  I join the fleet, then I add and configure the svc mesh.  It should be a serial action and I do it inline in the module when I create the cluster (the cluster create resource is omitted below).  It works for cluster-1, but fails on cluster-2 because (I THINK) the provider call isn't segmenting the action by membership.

```
resource "google_gke_hub_membership" "membership" {
  count = var.join_fleet ? 1 : 0
  project = var.project_id
  membership_id = google_container_cluster.cluster.name
  endpoint {
    gke_cluster {
      resource_link = google_container_cluster.cluster.id
    }
  }
  authority {
    issuer = "https://container.googleapis.com/v1/${google_container_cluster.cluster.id}"
  }
}

resource "google_gke_hub_feature_membership" "mesh" {
  count = var.join_fleet ? 1 : 0
  project = var.project_id
  location = "global"
  feature = "servicemesh"
  membership = google_gke_hub_membership.membership[0].id
  mesh {
    management = "MANAGEMENT_AUTOMATIC"
  }
  provider = google-beta
}

**Comment by djazayeri** [comment]
I am a complete terraform newbie, so maybe this is misguided, but could the issue be that  you need
```
    membership = google_gke_hub_membership.membership[${count.index}].id
```

**Comment by goobysnack** [comment]
> I am a complete terraform newbie, so maybe this is misguided, but could the issue be that you need
> 
> ```
>     membership = google_gke_hub_membership.membership[${count.index}].id
> ```

This is the state:
```
module.gcp_gkeclusters["cluster-1"].google_gke_hub_membership.membership[0]
module.gcp_gkeclusters["cluster-2"].google_gke_hub_membership.membership[0]
```
This is because the resource that creates the membership (IE join_fleet?) also uses count, and the value of the index is `0`

```
resource "google_gke_hub_membership" "membership" {
  count = var.join_fleet ? 1 : 0
  project = var.project_id
  membership_id = google_container_cluster.cluster.name
  endpoint {
    gke_cluster {
      resource_link = google_container_cluster.cluster.id
    }
  }
  authority {
    issuer = "https://container.googleapis.com/v1/${google_container_cluster.cluster.id}"
  }
}

**Comment by sahsagar-google** [comment]
@goobysnack Could you please try with just the id of the membership instead of the whole name/resource path. I have my suspicion that this could be the problem. Let me know if that be the case.

_Please see the changes inline:_ 
Terraform will perform the following actions:

  # module.gcp_gkeclusters["cluster-1"].google_gke_hub_feature_membership.mesh[0] will be created
  + resource "google_gke_hub_feature_membership" "mesh" {
      + feature    = "servicemesh"
      + id         = (known after apply)
      + location   = "global"
      + membership = _**"projects/project-1/locations/global/memberships/cluster-1" --> "cluster-1"**_
      + project    = "project-1"

      + mesh {
          + management = "MANAGEMENT_AUTOMATIC"
        }
    }

  # module.gcp_gkeclusters["cluster-2"].google_gke_hub_feature_membership.mesh[0] will be created
  + resource "google_gke_hub_feature_membership" "mesh" {
      + feature    = "servicemesh"
      + id         = (known after apply)
      + location   = "global"
      + membership = **_"projects/project-1/locations/global/memberships/cluster-2" --> "cluster-2"_**
      + project    = "project-1"

      + mesh {
          + management = "MANAGEMENT_AUTOMATIC"
        }
    }

**Comment by sahsagar-google** [comment]
From the documentation: https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/gke_hub_feature_membership#example-usage---config-management


`resource "google_container_cluster" "cluster" {
  name               = "my-cluster"
  location           = "us-central1-a"
  initial_node_count = 1
  provider = google-beta
}

resource "google_gke_hub_membership" "membership" {
  membership_id = "my-membership"
  endpoint {
    gke_cluster {
      resource_link = "//container.googleapis.com/${google_container_cluster.cluster.id}"
    }
  }
  provider = google-beta
}

resource "google_gke_hub_feature" "feature" {
  name = "configmanagement"
  location = "global"

  labels = {
    foo = "bar"
  }
  provider = google-beta
}

resource "google_gke_hub_feature_membership" "feature_member" {
  location = "global"
  feature = google_gke_hub_feature.feature.name
  membership = google_gke_hub_membership.membership.membership_id
  configmanagement {
    version = "1.6.2"
    config_sync {
      git {
        sync_repo = "https://github.com/hashicorp/terraform"
      }
    }
  }
  provider = google-beta
}`

**Comment by goobysnack** [comment]
I will try that, but I did base it on what is in the state file.

```
Terraform state show 'module.gcp_gkeclusters["cluster-1"].google_gke_hub_feature_membership.feature_member[0]'

# module.gcp_gkeclusters["cluster-1"].google_gke_hub_feature_membership.feature_member[0]:
resource "google_gke_hub_feature_membership" "feature_member" {
    feature    = "configmanagement"
    id         = "projects/<projectID>/locations/global/features/configmanagement/membershipId/<clustername>"
    location   = "global"
    membership = "projects/<projectID>/locations/global/memberships/<clustername>"
    project    = "<projectID>"

    configmanagement {
        version = "1.15.3"

        config_sync {
<removed>

            git {
<removed>
            }
        }

        policy_controller {
<removed>
                ]
            }
        }
    }
}
```

**Comment by goobysnack** [comment]
OH!  I was using 'membership' not 'id'!!

**Comment by goobysnack** [comment]
let me try 

`"projects/<projectID>/locations/global/features/configmanagement/membershipId/<clustername>"` instead of 

`"projects/<projectID>/locations/global/memberships/<clustername>"`

**Comment by goobysnack** [comment]
That was it!  Fixed!  Thanks for the help.

**Comment by edwardmedia** [comment]
closing as fixed
